New GM runtime parameter to allow in/excluding dynamically computed Kappa contributions in Redi-tensor

mjlosch opened the pull request:
## What changes does this PR introduce?

1. fix bug (internal flag `GM_ExtraDiag` not turned on when it should).
2. New feature

## What is the current behaviour? 

1. `GM_ExtraDiag` was not (always) turned on with 3D Kappa read from file, with Bates-K3d or QG-Leith, therefore missing two  terms in isopycnal (Redi) diffusion tensor. 
2. The dynamically computed variable kappa contribution of the Visbeck, etc. parameterisations is hardwired to be included in the Redi-tensor (see discussion in PR #674), although it is not always clear, if this intended.

## What is the new behaviour
1. `GM_ExtraDiag` set to True when needed.
2. A new runtime parameter `GM_isoFac_calcK` (`=1` by default) can be used to turn off the variable kappa contribution to the Redi-tensor. 

In my humble opinion, the new runtime parameter makes the code even a little easier to read, because I found the `GM_advect` parameter adding a contribution the Redi-tensor, when we do not use the skew flux form, a little confusing. Now it is replaced by `(GM_isoFac_calcK - GM_skewflx)` which makes it clear that the terms can cancel.

## Does this PR introduce a breaking change? 
Results change as consequence of the bug beeing fixed.
This affects test experiment using QG-Leith in GM (`MLAdjust.QGLthGM`) and the 2 test experiments in `verification_other`
that use Bates-K3d scheme. Seems that no current test exp. does read-in 3D Kappa from file. 

## Other information:
The code changes for the new runtime parameter are based on insights by @nicolasdettling.
These changes may have an effect on PR #641

## Suggested addition to `tag-index`
o pkg/gmredi:
  - fix setting of GM_ExtraDiag that was not (always) turned on with 3D Kappa
    read from file, with Bates-K3d or QG-Leith, therefore missing 2 terms in
    isopycnal diffusion tensor; update ref. output of MLAdjust.QGLthGM test ;
  - add more checks and stops for the case #undef GM_NON_UNITY_DIAGONAL;
  - add runtime parameter GM_isoFac_calcK to allow to scale dynamically computed
    Kappa contribution in the Redi-tensor (default = 1.).

jm-c commented:
@mjlosch I took a look at this PR and I think it's a valuable addition to enable this switch.
I have few comments regarding variable/parameter names and settings:

1. I could think of removing logical param "GM_useVarKforRedi" and put dirrectly "GM_varKredi" in namelist
  so that I could set it to, e.g., just half if needed.
2. If we keep the logical run-time param "GM_useVarKforRedi" , I would be in favor of renaming it to "GM_useVarKinRedi" since it's just one of many contribution to isopycnal diffusivity.
3. regarding the name "GM_varKredi": (a) it's a bit sad that "Redi" is not capitalized, but it's a detail ; (b) "var" (for variable) might leads to confusion with fixed in time, spacially varying K. We had issues before with CPP options names when setting 3-D kappa from files. (c) since it's a scaling factor (no units), could put "Fac" in the name.

I did not think too much about an alternative name for "GM_varKredi" but we might consider something like "GM_isoFac_calcK" (similar to "GM_isoFac2d" & "GM_isoFac1d") or a bit longer, "GM_isopycFac_calcK" (similar to "GM_isopycK"). Also we could think of introducing different scaling factor for the different schemes, like "GM_isoFac_Visbeck", "GM_isoFac_Bates" and "GM_isoFac_LeithQG" but this might not be really useful since the likelyhood to have 2 of these schemes turned on is low.

mjlosch commented:
@jm-c thanks for looking at this. I am fine with the name `GM_isoFac_calcK`, and I can also live with putting this directly into the namelist and thus getting rid of `GM_useVarKforRedi`.
I also though of having different parameters for the different scheme, but then I thought (as you did), it's unlikely that we will use more than one of the schemes at a time.

I can make this change tomorrow (unless you want to beat me to it tonight).

jm-c commented:
@mjlosch I am not going to make changes here soon, and it's also nice to wait until tomorrow in case someone want to suggest a different name.

jm-c reviewed (changes requested):
I did not check carefully, but 2 things might be missing:

1.  in `gmredi_calc_tensor.F` I don't see the new factor `GM_isoFac_calcK` inside the `Kwz` coeff, so I think it's missing there (around lines 591-601).
2. the setting of `GM_ExtraDiag` in `gmredi_readparams.F` needs to be updated for the skew-flux form (previously, there was an exact cancellation from calcK contributions).
3. a detail: I think `GM_advect` parameter is no longer used so it could be removed.

mjlosch commented:
> * in `gmredi_calc_tensor.F` I don't see the new factor `GM_isoFac_calcK` inside the `Kwz` coeff, so I think it's missing there (around lines 591-601).

Thanks for catching this. I fixed this in [32d769f](https://github.com/MITgcm/MITgcm/pull/769/commits/32d769f0df8d392292768d3523843321e820c219). 
I think it is OK to leave the
```
#ifdef GM_READ_K3D_GM
     &           + GM_skewflx*op5*( GM_inpK3dGM(i,j,km1,bi,bj)
     &                            + GM_inpK3dGM(i,j,k,bi,bj) )
#else
     &           + GM_skewflx*bolus_K*GM_bolFac2d(i,j,bi,bj)
#endif
```
without any factor.

> * the setting of `GM_ExtraDiag` in `gmredi_readparams.F` needs to be updated for the skew-flux form (previously, there was an exact cancellation from calcK contributions).

OK, see if I got it right with the latest commit [32d769f](https://github.com/MITgcm/MITgcm/pull/769/commits/32d769f0df8d392292768d3523843321e820c219)

> * a detail: I think `GM_advect` parameter is no longer used so it could be removed.

I agree, so far it is used in gmredi_check.F, from which it is now removed. See my modification.

jm-c reviewed (commented):
The changes in `gmredi_calc_tensor.F` look good now.
And regarding the setting of `GM_ExtraDiag`, it's better but also not completely right, I think, from even before adding `GM_isoFac_calcK` so I am going to check more carefully

jm-c commented on pkg/gmredi/GMREDI.h:
```diff
@@ -34,6 +35,7 @@ C     GM_useLeithQG    :: add Leith QG viscosity to GMRedi tensor
       LOGICAL GM_InMomAsStress
       LOGICAL GM_MNC
       LOGICAL GM_MDSIO
+      LOGICAL GM_useVarKforRedi
```
I think this new parameter can be removed

jm-c commented:
@mjlosch Thanks for the cleaning. 
I would  like to fix the setting of GM_ExtraDiag for  GM_useBatesK3d and GM_useLeithQG. This is not directly related to `GM_isoFac_calcK` addition (this PR) but depends on this new factor ; so easier to fix this here.  Do you agree ?

jm-c commented:
@mjlosch I think the setting of internal params "GM_ExtraDiag" is now fixed (please take a look and check) and as a results, the test experiment `MLAdjust.QGLthGM` is failing. Also, in `verification_other` the 2 global_ocean.gm_3d & gm_res test are failing too. Since I did not checked the OpenAD test, I added the label "adjoint" to this PR.

I've adjusted a little bit the parameter summary (printed from `gmredi_check.F`) and brought back the report of "GM_advec\*K ="  while updating the report of `GM_background_K` ; but I think we could keep only 1 of the 2, either "GM_advec\*K" or `GM_background_K`. I personally like the pair "GM_advec\*K" and "GM_skewflx\*K" but could change this.

mjlosch commented on pkg/gmredi/gmredi_readparms.F:
```diff
@@ -205,20 +207,40 @@ SUBROUTINE GMREDI_READPARMS( myThid )
-        GM_advect  = 0.
-        GM_ExtraDiag = GM_isopycK.NE.GM_background_K
+        GM_skewflx = 1. _d 0
+        IF ( GM_K3dRediFile .NE. ' ' .OR.
+     &       GM_K3dGMFile   .NE. ' ' ) THEN
```
I have the impression that in analogy to lines 234--235, this should also read
```suggestion
        IF ( ( GM_K3dRediFile .NE. ' ' .OR.
     &         GM_K3dGMFile   .NE. ' ' )  .AND.
     &       ( GM_K3dRediFile .NE. GM_K3dGMFile  ) ) THEN
```

mjlosch replied:
or just
```
      IF ( GM_K3dRediFile .NE. GM_K3dGMFile  ) THEN
```
?

mjlosch commented on pkg/gmredi/gmredi_readparms.F:
```diff
@@ -205,20 +207,40 @@ SUBROUTINE GMREDI_READPARMS( myThid )
+        IF ( GM_K3dRediFile .NE. ' ' .OR.
+     &       GM_K3dGMFile   .NE. ' ' ) THEN
+          GM_ExtraDiag = .TRUE.
+        ELSEIF ( GM_iso2dFile .NE. GM_bol2dFile .OR.
+     &           GM_iso1dFile .NE. GM_bol1dFile ) THEN
```
If all files are `' '`, then this is true, isn't it? So I suggest:
```suggestion
        ELSEIF ( ( GM_iso2dFile .NE. ' ' .OR. GM_bol2dFile .NE. ' ' .OR.
     &             GM_iso1dFile .NE. ' ' .OR. GM_bol1dFile .NE. ' ') .AND.
     &           ( GM_iso2dFile .NE. GM_bol2dFile .OR.
     &             GM_iso1dFile .NE. GM_bol1dFile ) ) THEN
```

mjlosch commented on pkg/gmredi/gmredi_check.F:
```diff
@@ -76,12 +78,19 @@ SUBROUTINE GMREDI_CHECK( myThid )
+       CALL WRITE_0D_RL( GM_background_K, INDEX_NONE,
+     &  'GM_background_K =',
+     &  ' /* Background GM (=Bolus) Diffusivity [m^2/s] */')
+       tmpVar = GM_background_K*( oneRL - GM_skewflx )
+       CALL WRITE_0D_RL( tmpVar, INDEX_NONE, 'GM_advec*K =',
+     &  '    /* Backg. GM-Advec(=Bolus) Diffusivity [m^2/s] */')
```
I think it's a good idea to print `tmpVar`, but now that `GM_advec` does not exist anymore, I find `GM_advec*K` a bit misleading. Since I do not have a better idea for a name here, I am OK with this

jm-c replied:
Also, I modified the report of `GM_isoFac_calcK` description (lines 91-93) keeping your version commented out. If you preferer to keep your version it's fine and it's easy to change back.

mjlosch replied:
I am fine with whatever you find useful. Maybe the current version of the file is the best (including the description of `GM_isoFac_calcK`.

mjlosch reviewed (commented):
This quickly goes over my head, especially the if-block frenzy in gmredi_readparms.F, see my inline comments. I am not sure if I got it right.

jm-c commented:
@mjlosch I cleaned up a little bit (mostly the changes I made earlier) and updated one test exp. ref. output (the one using QGLeith in GM).
I think this PR is getting ready to be merged, unless you want to make some changes. 
Regarding this PR description, do you want to add something about the bug fixed or do you prefer that I do it ?

mjlosch commented:
@jm-c LGTM, please update the description, as you know better what you have done.

jm-c commented:
@mjlosch I updated this PR description (+ suggested addition to tag-index). Please review and change as you like.
After this, we could merge this PR soon (like tomorrow) unless someone wants to add something.

jm-c commented:
@mjlosch I have a follow up question regarding ALLOW_KAPREDI_CONTROL and ALLOW_KAPGM_CONTROL:
Do we need to set `GM_ExtraDiag` (to True) if one of these is defined ?
I did that for the `global_oce_latlon` TAF test (where `GM_ExtraDiag` was not set) but did not see any differences there.
And I forgot what the `grdchkvarindex = 201` is testing.

jm-c commented:
@mjlosch Update reagarding this:
> I did that for the global_oce_latlon TAF test (where GM_ExtraDiag was not set) but did not see any differences there.

The STDOUT are unchanged (with grdchkvarindex = 201) but binary files "ADJkapgm" and "adxx_kapgm" are different.
So at the end, I pushed my (last) changes to `gmredi_readparms.F`.

mjlosch commented:
According to `s/r ctrl_check`, `ALLOW_KAPREDI_CONTROL` and `ALLOW_KAPGM_CONTROL` require `GM_READ_K3D_REDI` and `GM_READ_K3D_GM`, but I guess `GM_K3dRedi/GMFile` can be empty, so your changes look good. I changed the PR descriptions slightly (mostly because it now contains more than one change).

I think that this is ready to be merged.

PS. I never know about they `grdchkvarindex`, after inspecting the stdout, I usually come to the conclusion that it refers to the reported `index` (so `xx_theta` in the case of `global_oce_latlon`):
```
(PID.TID 0000.0001)  -> 3d control, genarr3d no.  1 is in use
(PID.TID 0000.0001)       file       = xx_theta
(PID.TID 0000.0001)       weight     = ones_64b.bin
(PID.TID 0000.0001)       index      =  0201     <<<<<<<<<<<<USE THIS FOR GRDCHKVARINDEX<<<<<<<<<
(PID.TID 0000.0001)       ncvarindex =  0301
(PID.TID 0000.0001)  -> 3d control, genarr3d no.  2 is in use
(PID.TID 0000.0001)       file       = xx_salt
(PID.TID 0000.0001)       weight     = ones_64b.bin
(PID.TID 0000.0001)       index      =  0202
(PID.TID 0000.0001)       ncvarindex =  0302
(PID.TID 0000.0001)  -> 3d control, genarr3d no.  3 is in use
(PID.TID 0000.0001)       file       = xx_diffkr
(PID.TID 0000.0001)       weight     = ones_64b.bin
(PID.TID 0000.0001)       index      =  0203
(PID.TID 0000.0001)       ncvarindex =  0303
(PID.TID 0000.0001)  -> 3d control, genarr3d no.  4 is in use
(PID.TID 0000.0001)       file       = xx_kapgm
(PID.TID 0000.0001)       weight     = ones_64b.bin
(PID.TID 0000.0001)       index      =  0204
(PID.TID 0000.0001)       ncvarindex =  0304
```
it's also confusing that `index` and `ncvarindex` are not the same ... oh, well. Maybe we should change `index` to `grdchkvarindex` in the report? (not in this PR).

jm-c commented:
@mjlosch Thanks for the comments. I will merge this PR later today.
